//BookStore.proto

syntax = "proto3";

service BookStore {
  rpc  access_to_server(AccessRequest) returns (AccessResponse) {}
  rpc LocalStorePS(LocalStorePSRequest) returns (LocalStorePSResponse) {}
  rpc CreateChain(CreateChainRequest) returns (CreateChainResponse) {}
  rpc ListChain(ListChainRequest) returns (ListChainResponse) {}
  rpc WriteOperation(WriteOperationRequest) returns (WriteOperationResponse) {}
  rpc ListBooks(ListBooksRequest) returns (ListBooksResponse) {}
  rpc ReadOperation(ReadOperationRequest) returns (ReadOperationResponse) {}
  rpc Timeout(TimeoutRequest) returns (TimeoutResponse) {}
  rpc DataStatus(DataStatusRequest) returns (DataStatusResponse) {}
  rpc RemoveHead(RemoveHeadRequest) returns (RemoveHeadResponse) {}
  rpc RestoreHead(RestoreHeadRequest) returns (RestoreHeadResponse) {}
}

message AccessRequest {
  string name = 1 ;
}

message AccessResponse {
  string id = 1;
}

message LocalStorePSRequest {
  int32 k = 1;
}

message LocalStorePSResponse {
  repeated string process_ids = 1;
}

message CreateChainRequest {}

message CreateChainResponse {
  repeated string process_ids = 1;
}

message ListChainRequest {}

message ListChainResponse {
  string chain = 1;
}

message WriteOperationRequest {
  string book = 1;
  double price = 2;
}

message WriteOperationResponse {}

message ListBooksRequest {}

message ListBooksResponse {
  repeated string books = 1;
}

message ReadOperationRequest {
  string book = 1;
}

message ReadOperationResponse {
  double price = 1;
  bool exists = 2;
}

message TimeoutRequest {
  int32 timeout = 1;
}

message TimeoutResponse {}

message DataStatusRequest {}

message DataStatusResponse {
  repeated string statuses = 1;
}

message RemoveHeadRequest {}

message RemoveHeadResponse {
  string new_head = 1;
}

message RestoreHeadRequest {}

message RestoreHeadResponse {
  string new_head = 1;
}
